Fitbit.Portable/FitBitHttpClientMessageHandler.cs

```diff
@@ -1,38 +1,59 @@
 ﻿namespace Fitbit.Api.Portable
 {
+    using System;
     using System.Diagnostics;
     using System.Net.Http;
     using System.Threading;
     using System.Threading.Tasks;
 
-    public class FitbitHttpClientMessageHandler : DelegatingHandler
+    internal class FitbitHttpClientMessageHandler : DelegatingHandler
     {
-        private IFitbitClientInterceptor logger;
+        private IFitbitInterceptor interceptor;
+        Func<Task<HttpResponseMessage>, CancellationToken, Task<HttpResponseMessage>> responseHandler;
 
-        public FitbitHttpClientMessageHandler(IFitbitClientInterceptor logger)
+        //private ITokenManager tokenManager;
+
+        public FitbitHttpClientMessageHandler(IFitbitInterceptor interceptor)
         {
-            this.logger = logger;
+            this.interceptor = interceptor;
+            //this.tokenManager = tokenManager;
+            responseHandler = ResponseHandler; 
             //Define the inner must handler. Otherwise exception is thrown.
             InnerHandler = new HttpClientHandler();
         }
 
+        //Handle the following method with EXTREME care as it will be invoked on ALL requests made by FitbitClient
         // We override the SendAsync method to intercept both the request and response path
         protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
+            Task<HttpResponseMessage> interceptorResponse = null;
             Debug.WriteLine("Entering Http client's request message handler. Request details: {0}", request.ToString());
-            if (logger != null)
-                logger.InterceptRequest(request, cancellationToken);
-
-            return base.SendAsync(request, cancellationToken).ContinueWith(
-                     requestTask =>
-                     {
-                         Debug.WriteLine("Entering Http client's response message handler. Response details: {0}", requestTask.Result);
-                         if (logger != null)
-                             logger.InterceptResponse(requestTask.Result, cancellationToken);
-
-                         return requestTask.Result;
-                     }, TaskContinuationOptions.OnlyOnRanToCompletion
+
+            if (interceptor != null)
+                interceptorResponse = interceptor.InterceptRequest(request, cancellationToken);
+
+            if(interceptorResponse != null) //then highjack the request pipeline and return the HttpResponse returned by interceptor. Invoke Response handler at return.
```

Comment:
@WestDiscGolf The param [] seems a little overkill. For the 1% of folks who may want to wedge code in between request / response, they can do that in a single plan and then route multiple methods in there. We went back and fourth over here at the desks and Mau had a good point that tests are very easy to write when you can intercept the HTTP pipeline. In our plans, we want to log all HTTP call / responses, and we also might want to put a counter or something like that, which we can put in a single codeblock. Fitbit's support folks will probably like this feature here because so often people post code in the forums where to debug they'd really just like HTTP traffic!